pkg/s3-proxy/server/tls.go

```diff
@@ -0,0 +1,228 @@
+	case certConfig.Certificate != nil:
+		certificate = []byte(*certConfig.Certificate)
+
+	// Certificate supplied as a URL.
+	case certConfig.CertificateURL != nil:
+		var err error
+		certificate, err = utils.GetDocumentFromURL(*certConfig.CertificateURL)
+
+		if err != nil {
+			logger.Fatalf("Failed to get certificate from URL %s: %v", *certConfig.CertificateURL, err)
+		}
+
+	default:
+		logger.Fatal("Expected either certificate or certificateUrl to be set")
+	}
+
+	switch {
+	// Private key supplied directly; just copy it.
+	case certConfig.PrivateKey != nil:
+		privateKey = []byte(*certConfig.PrivateKey)
+
+	// Private key supplied as a URL.
+	case certConfig.PrivateKeyURL != nil:
+		var err error
+		privateKey, err = utils.GetDocumentFromURL(*certConfig.PrivateKeyURL)
+
+		if err != nil {
+			logger.Fatalf("Failed to get certificate private key from URL %s: %v", *certConfig.PrivateKeyURL, err)
+		}
+
+	default:
+		logger.Fatal("Expected either privateKey or privateKeyUrl to be set")
+	}
+
+	cert, err := tls.X509KeyPair(certificate, privateKey)
+
+	if err != nil {
+		return tls.Certificate{}, fmt.Errorf("failed to create certificate: %w", err)
+	}
+
+	if len(cert.Certificate) == 0 {
+		return tls.Certificate{}, errors.New("no certificates loaded")
+	}
+
+	for _, cert := range cert.Certificate {
+		if len(cert) == 0 {
+			return tls.Certificate{}, errors.New("empty certificate loaded")
+		}
+	}
+
+	return cert, nil
+}
+
+// generateSelfSignedCertificate returns a single crypto/tls.Certificate containing a self-signed certficate for
+// the specified hostnames.
+func generateSelfSignedCertificate(hostnames []string) (tls.Certificate, error) {
+	privateKey, err := rsa.GenerateKey(rand.Reader, rsaKeySize)
+
+	if err != nil {
+		return tls.Certificate{}, fmt.Errorf("failed to generate RSA key: %w", err)
+	}
+
+	now := time.Now().UTC()
+
+	// Make the start time an hour earlier to account for clock skew.
+	startTime := now.Add(-1 * time.Hour)
+
+	// End time is 10 years.
+	endTime := startTime.Add(certValidityDuration)
+
+	// Generate a universally unique serial number.
+	one := big.NewInt(1)
+	maxSerialNumber := &big.Int{}
+	maxSerialNumber.Lsh(one, certSerialBits)
+	serialNumber, err := rand.Int(rand.Reader, maxSerialNumber)
+
+	if err != nil {
+		return tls.Certificate{}, fmt.Errorf("failed to generate serial number: %w", err)
+	}
+
+	template := x509.Certificate{
+		DNSNames:           hostnames,
+		KeyUsage:           x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
+		NotAfter:           endTime,
+		NotBefore:          startTime,
+		SerialNumber:       serialNumber,
+		SignatureAlgorithm: x509.SHA256WithRSA,
+		Subject:            pkix.Name{CommonName: hostnames[0]},
+	}
+
+	certDER, err := x509.CreateCertificate(rand.Reader, &template, &template, privateKey.Public(), privateKey)
+
+	if err != nil {
+		return tls.Certificate{}, fmt.Errorf("failed to create self-signed certificate: %w", err)
```

Comment:
Can you wrap the fmt with a errors.WithStack?

Reply:
The `%w` formatting directive should be the same (might produce slightly better results) as `errors.WithStack`: https://go.dev/blog/go1.13-errors

Should I change `%w` to `%v` and use `errors.WithStack` (allows for pre 1.13 compatibility) still?

Reply:
The `%w` will keep the stack but the created error on line 217 with the Errorf won't have it. I must admit that I never used the `%w` before. I think you can keep it and wrap just it, the logger should manage this. Thanks!

Reply:
The `%w` formatting directive should be the same (might produce slightly better results) as `errors.WithStack`: https://go.dev/blog/go1.13-errors

Should I change `%w` to `%v` and use `errors.WithStack` (allows for pre 1.13 compatibility) still?

Reply:
Argh, sorry for the repeated comments. I'm on airplane WiFi and it's not behaving well with GitHub.

Reply:
Sure, no problem :).